src/caches/beliefpropagationcache.jl

```diff
@@ -0,0 +1,221 @@
+  ptn = partitioned_itensornetwork(bpc)
+  pb_edges = partitionedges(ptn, boundary_edges(ptn, vertices(ptn, src(edge)); dir=:in))
+
+  incoming_messages = [
+    message_tensor(bpc, e_in; mt_constructor) for e_in in setdiff(pb_edges, [reverse(edge)])
+  ]
+  incoming_messages = reduce(vcat, incoming_messages; init=ITensor[])
+
+  contract_list = ITensor[
+    incoming_messages
+    Vector{ITensor}(subgraph(ptn, src(edge)))
+  ]
+
+  return contractor(contract_list; contractor_kwargs...)
+end
+
+"""
+Do a sequential update of the message tensors on `edges`
+"""
+function update(
+  bpc::BeliefPropagationCache,
+  edges::Vector{<:PartitionEdge};
+  compute_norm=false,
+  mt_constructor=default_mt_constructor,
+  kwargs...,
+)
+  bpc_updated = copy(bpc)
+  mts = message_tensors(bpc_updated)
+  c = 0
+  for e in edges
+    mts[e] = updated_message_tensor(bpc_updated, e; mt_constructor, kwargs...)
+    if compute_norm
+      c += message_tensor_diff(message_tensor(bpc, e; mt_constructor), mts[e])
+    end
+  end
+  return bpc_updated, c / (length(edges))
+end
+
+"""
+Update the message tensor on a single edge
+"""
+function update(bpc::BeliefPropagationCache, edge::PartitionEdge; kwargs...)
+  return update(bpc, [edge]; kwargs...)
+end
+
+"""
+Do parallel updates between groups of edges of all message tensors
+Currently we send the full message tensor data struct to update for each edge_group. But really we only need the
+mts relevant to that group.
+"""
+function update(
+  bpc::BeliefPropagationCache, edge_groups::Vector{<:Vector{<:PartitionEdge}}; kwargs...
+)
+  new_mts = copy(message_tensors(bpc))
+  c = 0
+  for edges in edge_groups
+    bpc_t, ct = update(bpc, edges; kwargs...)
+    for e in edges
+      new_mts[e] = message_tensor(bpc_t, e)
+    end
+    c += ct
+  end
+  return BeliefPropagationCache(copy(partitioned_itensornetwork(bpc)), new_mts),
+  c / (length(edge_groups))
+end
+
+"""
+More generic interface for update, with default params
+"""
+function update(
+  bpc::BeliefPropagationCache;
+  edges=default_edge_sequence(bpc),
+  niters=default_bp_niters(bpc),
+  target_precision=nothing,
+  verbose=false,
+  kwargs...,
+)
+  compute_norm = !isnothing(target_precision)
+  if isnothing(niters)
+    error("You need to specify a number of iterations for BP!")
+  end
+  for i in 1:niters
+    bpc, c = update(bpc, edges; compute_norm, kwargs...)
+    if compute_norm && c <= target_precision
+      if verbose
+        println("BP converged to desired precision after $i iterations.")
+      end
+      break
+    end
+  end
+  return bpc
+end
+
+"""
+Update the tensornetwork inside the cache
+"""
+function update(bpc::BeliefPropagationCache, states::Vector{ITensor}, vertices::Vector)
+  bpc = copy(bpc)
+  tn = tensornetwork(bpc)
+
+  for (state, vertex) in zip(states, vertices)
+    setindex_preserve_graph!(tn, state, vertex)
+  end
+  return bpc
+end
+
+function update(bpc, state, vertex)
+  return update(bpc, ITensor[state], [vertex])
+end
```

Comment:
I think we should come up with a different name than `update`, since that is used for updating messages throughout the rest of the code and I think this function could easily be confused for that functionality.

Maybe we could use `update_factors` to use BP terminology. Also, I think it is better to order the arguments as `update_factors(bp_cache::BeliefPropagationCache, vertices, factors)` instead of `update_factors(bp_cache::BeliefPropagationCache, factors, vertices)`, i.e. pass the vertices and then the factors.

Reply:
Also, I think we could get rid of the version taking one vertex, since it's not clear how to distinguish between a vector of vertices vs. a vertex that uses a vector as the name.

Reply:
`setindex!` in Julia has a conventional ordering of `setindex!(object, value, index)`, so I try to follow that argument ordering for `setindex!` and related functions, but with other functions I prefer the ordering `f(object, index, value)` since that is more natural to me, and is also followed by other functions in the ecosystem ([setfield!](https://docs.julialang.org/en/v1/base/base/#Core.setfield!),  [Dictionaries.set!](https://github.com/andyferris/Dictionaries.jl), [Accessors.set](https://juliaobjects.github.io/Accessors.jl/stable/docstrings/#Accessors.set), etc.).

Reply:
Okay I see, I will go with `index, value` then for `update_factors`